### PR TITLE
feat(indexer): Consumer builder using routing strategy

### PR DESCRIPTION
**Pull request by nikhars**
### Context
In a world where a specific dataset of Snuba is sliced, the Snuba consumers work on sliced topics. That means Snuba consumers belonging to separate slices consume from separate topics. In order to support that, the input to the Snuba consumers i.e. metrics indexer needs to have the capability to write to different topics. This can be achieved by using the RoutingProducer introduced in https://github.com/getsentry/sentry/pull/40776

### Implementation
Since the strategies can get opened/closed whenever Kafka rebalance happens, the producer resources need to be created before the consumer builder factory and gets passed into it. The `reconstruct_message` in IndexerBatch is modified to either return normal kafka payload or routing payload based on whether it needs to send output specifically for routing or not.

### Tests
Added a unit test for the Indexer changes to correctly reconstruct messages for the sliced output.
Ran the consumer locally with the settings override to confirm that messages get sent to the new sliced topic

### Performance test results
Baseline test - 18k msgs/second
Screen Shot 2022-12-09 at 10 51 35 PM

With slicing enabled - 15k msgs/second
Screen Shot 2022-12-09 at 10 51 26 PM

There is a degradation of 3k msgs/second which I am not too concerned about right now since slicing isn't needed right now

### CODE REVIEW

File: requirements-base.txt [hunk 1]
```diff
@@ -54,7 +54,7 @@ requests>=2.25.1
 rfc3339-validator>=0.1.2
 rfc3986-validator>=0.1.1
 # [end] jsonschema format validators
-sentry-arroyo>=2.1.0
+git+https://github.com/getsentry/arroyo@2.2.0#egg=sentry-arroyo
```
Comment (evanh) [inline requirements-base.txt:57]: ?

Comment (untitaker) [reply requirements-base.txt:57]: sentry-arroyo 2.2.0 is out, you no longer need to do this. also https://github.com/getsentry/sentry/pull/42218#pullrequestreview-1212012681 will bump arroyo

Comment (asottile-sentry) [reply requirements-base.txt:57]: weird, the linter should have also **prevented** this, something isn't right about that 🤔

Comment (nikhars) [reply requirements-base.txt:57]: This was primarily added to do performance runs with a package which had not been released yet. Now that it has been released, I will revert to the old way

Comment (nikhars) [reply requirements-base.txt:57]: > weird, the linter should have also **prevented** this, something isn't right about that 🤔

Well there are cases when you want to build against a commit or tag which hasn't been released yet. If the linter prevents it what would the alternative be for those use cases?

Comment (asottile-sentry) [reply requirements-base.txt]: it would still allow your CI run to happen -- but it would not allow it to be merged in this state

**Review by onewland — APPROVED** [review]
looks good to me

**Pull request merged**